crates/sui-indexer-alt/src/handlers/sum_obj_types.rs

```diff
@@ -0,0 +1,195 @@
+                        }));
+                    }
+                }
+            }
+
+            // Modified and created objects.
+            for object in &tx.output_objects {
+                let object_id = object.id().to_vec();
+                let object_version = object.version().value() as i64;
+                match values.entry(object_id.clone()) {
+                    Entry::Occupied(entry) => {
+                        ensure!(entry.get().object_version() > object_version);
+                    }
+
+                    Entry::Vacant(entry) => {
+                        let type_ = object.type_();
+                        entry.insert(Update::Update(StoredSumObjType {
+                            object_id,
+                            object_version,
+
+                            owner_kind: match object.owner() {
+                                Owner::AddressOwner(_) => StoredOwnerKind::Address,
+                                Owner::ObjectOwner(_) => StoredOwnerKind::Object,
+                                Owner::Shared { .. } => StoredOwnerKind::Shared,
+                                Owner::Immutable => StoredOwnerKind::Immutable,
+                            },
+
+                            owner_id: match object.owner() {
+                                Owner::AddressOwner(a) => Some(a.to_vec()),
+                                Owner::ObjectOwner(o) => Some(o.to_vec()),
+                                _ => None,
+                            },
+
+                            package: type_.map(|t| t.address().to_vec()),
+                            module: type_.map(|t| t.module().to_string()),
+                            name: type_.map(|t| t.name().to_string()),
+                            instantiation: type_
+                                .map(|t| bcs::to_bytes(&t.type_params()))
+                                .transpose()
+                                .map_err(|e| {
+                                    anyhow!(
+                                        "Failed to serialize type parameters for {}: {e}",
+                                        object.id().to_canonical_display(/* with_prefix */ true),
+                                    )
+                                })?,
+                        }));
+                    }
+                }
+            }
+        }
+
+        Ok(values.into_values().collect())
+    }
+}
+
+#[async_trait::async_trait]
+impl Handler for SumObjTypes {
+    type Batch = BTreeMap<Vec<u8>, Update>;
+
+    fn batch(batch: &mut Self::Batch, updates: Vec<Self::Value>) {
+        // `updates` are guaranteed to be provided in checkpoint order, so blindly inserting them
+        // will result in the batch containing the most up-to-date update for each object.
+        for update in updates {
+            batch.insert(update.object_id(), update);
+        }
+    }
+
+    async fn commit(values: &Self::Batch, conn: &mut db::Connection<'_>) -> anyhow::Result<usize> {
```

Comment:
can you remind me why you chose a `conn` instead of a `pool` here, as naively `try_join_all` will interleave on execution?

Reply:
The connection is tied to a transaction -- all these commits are done atomically, alongside the watermark update. This is required for consistency but something that the current indexer does not do.